### PR TITLE
Generalized OHEM

### LINKED ISSUE #52: About mmseg OHEMSampler Implementation
##### I noticed that maybe there are some problems of OHEMSampler Implementation in mmseg:
   - 1.  The following code will not work , no matter what `mask` and `sort_prob` are , the values in `seg_weight` will not change.
      ```python
           seg_weight[mask][sort_prob < threshold] = 0.
      ```

   - 2. Your OHEMSampler Implementation try to assign 0 weight to pixels which have low confidence. But these pixels always get high loss value in cross_entropy,maybe there are some logical probloms in the implementation?

   - I try to improve the implementation in following way : 
      ```python
               def sample(self, seg_logit, seg_label):
  
                     with torch.no_grad():
                            assert seg_logit.shape[2:] == seg_label.shape[2:]
                            assert seg_label.shape[1] == 1
                            seg_label = seg_label.squeeze(1).long()
                            batch_kept = self.min_kept * seg_label.size(0)
                            seg_prob = F.softmax(seg_logit, dim=1)
                            mask = seg_label.contiguous().view(-1, ) != self.ignore_index

                            tmp_seg_label = seg_label.clone()
                            tmp_seg_label[tmp_seg_label == self.ignore_index] = 0
                            seg_prob = seg_prob.gather(1, tmp_seg_label.unsqueeze(1))
                            sort_prob, sort_indices = seg_prob.contiguous().view(
                                     -1, )[mask].contiguous().sort()

                            if sort_prob.numel() > 0:
                                   min_threshold = sort_prob[min(batch_kept,
                                                    sort_prob.numel() - 1)]
                            else:
                                   min_threshold = 0.0
                            threshold = max(min_threshold, self.thresh)

                            seg_weight = seg_logit.new_ones(size=seg_label.size())
                            seg_weight = seg_weight.view(-1)
                            tmp = seg_weight[mask]
                            tmp[sort_prob > threshold] = 0
                            seg_weight[mask] = tmp
                            seg_weight = seg_weight.view_as(seg_label)

                           return seg_weight
      ```

   - I‘ll be really appreciated if you could give me some advices! @xvjiarui @hellock

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR make OHEM more generalized. 
It supports 2 modes: 
1. Select hard examples by threshold on prediction probability. 
2. Select hard examples by loss. 

It also fixed #52 .

### CODE REVIEW

File: mmseg/core/seg/sampler/ohem_pixel_sampler.py [hunk 1]
```diff
@@ -10,55 +10,67 @@ class OHEMPixelSampler(BasePixelSampler):
...
         super(OHEMPixelSampler, self).__init__()
+        self.context = context
         assert min_kept > 1
         self.thresh = thresh
         self.min_kept = min_kept
-        self.ignore_index = ignore_index
 
     def sample(self, seg_logit, seg_label):
-        """
+        """Hard examples sample.
```
Comment (hellock) [inline mmseg/core/seg/sampler/ohem_pixel_sampler.py:31]: Too informal.
